### ISSUE TITLE
Package yrs binaries as Nuget packages

### ISSUE THREAD

**Issue by goldsam** [LSViana/ydotnet#15]
Consumption of this work (upon release) would be easier if the Yrs binaries were provided as nuget packages. If you would like, I could put together a PR for this.

**Comment by LSViana** [comment]
Hi, @goldsam!

That would be definitely appreciated.

This is already part of the plan since I didn't plan to put the burden of providing the Yrs binaries on the clients anyway, but this would only be relevant after finishing the bindings.

**Comment by goldsam** [comment]
It seems like your making amazing progress very quickly. I'll see what I can do

**Comment by goldsam** [comment]
@LSViana I've made a [little bit of progress in generating and packaging the binaries](https://github.com/goldsam/ydotnet/tree/binaries). The more challenging part of this task is selection of the binary to load at runtime. I was going to take some inspiration from [SkiaSharp](https://github.com/mono/SkiaSharp/tree/main/binding/Binding.Shared) and [Tesseract](https://github.com/charlesw/tesseract/tree/master/src/Tesseract/Internal/InteropDotNet) \(the later of which uses [Interop.net](https://github.com/AndreyAkinshin/InteropDotNet) internally\) but wanted to first get your opinion before doing any work you might disagree with.

**Comment by goldsam** [comment]
Following your feedback from #19 , my updated build against main of LSViana/y-crdt shows 4 tests failing (165 passing). Is that expected?

**Comment by LSViana** [comment]
> @LSViana I've made a [little bit of progress in generating and packaging the binaries](https://github.com/goldsam/ydotnet/tree/binaries). (...)

That looks great to me! I see you were able to build for multiple platforms successfully and that's nice.

I think we can work on downloading the correct binaries for the running platform & architecture during the build step from the published artifacts, right? If the file is already downloaded, we just skip not to slow down every build.

I don't have lots of experience with that, to be honest, but what you started there is promising and if I can help with it after I'm done with the library bindings, I'll definitely try. 🙂

> @LSViana Following your feedback from https://github.com/LSViana/ydotnet/issues/19 , my updated build against main of LSViana/y-crdt shows 4 tests failing (165 passing). Is that expected?

I just noticed that some tests are failing on Windows and, no, it's not expected. I've been mixing my development time between Unix and Windows machines to ensure the bindings are working on both sides but I let these issues go unnoticed.

The issues with the `ObserveDeep*` tests have probably already been fixed by [this commit](https://github.com/LSViana/y-crdt/commit/c8adec24d4c23655a018befc34ee31e9fd370107) on my fork of Yrs. Please, can you check if you have that update locally?

The issues with emojis are probably due to the usage of UTF-16 characters because the C bindings definitions are ready for UTF-8 characters. I created those tests exactly to try edge cases and, well, I found them now. 😅

I don't have an immediate fix for these failing tests related to UTF-16 characters, feel free to disable them on your fork of the repository if they're blocking your GitHub Actions pipelines, for example.

**Comment by LSViana** [comment]
@goldsam An update here: the issues with emoji and other special characters should be fixed by now after the commits I pushed today. I tested them both on Windows and macOS and the tests are all passing now.

Beyond fixing these 2 failing tests, I actually applied a library-wide update to read and write `string` instances transported through the FFI to make sure we have full control instead of relying on the default mechanisms that can't handle emojis and other special characters.

If you have any issues with these latest updates, please let me know.

**Comment by SebastianStehle** [comment]
@goldsam are you still working on this? I would really like to use yjs in a .net project so I am pushing this and the server components. But I have almost zero experience with rust.

**Comment by goldsam** [comment]
@SebastianStehle Did you not see the link to my repo above?

**Comment by Skyppid** [comment]
Same here, been looking for Yjs compatible port for C# that supports the current Yjs featureset. I also never worked with Rust and it's really a turn off having to struggle with that. A pipeline for this repository that builds NuGet packages with the necessary libaries would be amazing :)

The pipeline looks good, but are there any packages? Or what's the status of it @goldsam?

**Comment by LSViana** [comment]
Hi, @Skyppid!

There's a PR that is supposed to fix that and should be merged soon really soon (see #55).

**Comment by Skyppid** [comment]
Ah I saw this one, but didn't realize that it also included the pipeline. Quite a huge PR.
Nonetheless I'm looking forward to see it merged! Definetly gonna try it out and help in case I find any issues. Gonna need this for a project I've been wanting to do for a long time now and it's the only missing puzzle piece 😄 

Thanks for the quick update :)

**Comment by LSViana** [comment]
Hi, @Skyppid!

PR #55 was just merged and you should be able to see some improvements to the strategy to include binaries for each platform in there now.

If you have questions, please let me know.

**Comment by Skyppid** [comment]
Amazing, thank you guys! I'll be checking it out as soon as I got time. I'll support wherever I can :)
Do I see it correctly that Ydotnet does not only provide the basic Y components but also some management classes already for .NET backends?

I'll use it with SignalR, maybe I can provide a basic framework for people to use it without much work on their own if @SebastianStehle isn't doing this already.

**Comment by SebastianStehle** [comment]
Hi, 
yes YDotnet provides a websocket backend now. But there is no clustering feature yet. I was playing around here: https://github.com/SebastianStehle/ydotnet/tree/main/YDotNet.Server/Clustering, but I never put it in the pr. 

The reason is that I don't have the opportunity to battle test it at the moment. The question where to handle persistency and so on is challenging because the server is stateful. In general I think the best solution would be to use something like actors or grains where exactly on server stores a single document (e.g. with Microsoft Orleans), but this has other challenges around cluster management.

So for my use case, where I have only very few people working with the yjs functionality I have decided to declare one node as yjs node. So I basically have N normal nodes and one yjs node (with a environment variable) and the UI connects to this dedicated node.

**Comment by Skyppid** [comment]
Alright, fair points. Well for me it's not necessary as of now. I'm only building a prototype for now so no need for clustering yet. Is the websocket implementation compatible with Yjs' websockets (looks like it from what I saw)?

**Comment by SebastianStehle** [comment]
Yes, it is. It uses the same encoders and stuff from y-protocols, 
 e.g. https://github.com/LSViana/ydotnet/blob/main/YDotNet.Server.WebSockets/EncoderExtensions.cs
 
 And in the demo I use the normal websocket provider:
 https://github.com/LSViana/ydotnet/blob/main/Demo/Client/src/context/yjsContext.tsx
 
 It does not implement auth yet, but the y-websocket provider also does not support it.

**Comment by Skyppid** [comment]
Well no auth is an issue, but also something I can ignore for now. Shouldn't be too hard to implement, though. Something like that: https://github.com/yjs/y-websocket/issues/7#issuecomment-623114183

**Comment by SebastianStehle** [comment]
No auth is not entirely correct. You can implement the normal auth, but I think some headers are not supported for web sockets and stuff like this, therefore it can make sense to implement in the websocket part directly and not outside.
